integration-tests/application-server-integration-tests/src/test/java/co/elastic/apm/servlet/ApmServerRule.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.servlet;
+
+import co.elastic.apm.agent.sdk.logging.Logger;
+import co.elastic.apm.agent.sdk.logging.LoggerFactory;
+import okhttp3.mockwebserver.Dispatcher;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import okio.Buffer;
+import okio.InflaterSource;
+import org.junit.rules.ExternalResource;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.testcontainers.Testcontainers;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.zip.Inflater;
+
+import static co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.INTAKE_V2_FLUSH_URL;
+import static co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.INTAKE_V2_URL;
+
+/**
+ * This is modeled after {@code co.elastic.apm.awslambda.fakeserver.FakeApmServer}, except
+ * deflate instead of gzip and OkHttp instead of {@code com.sun.net.httpserver.HttpServer}.
+ */
+public class ApmServerRule extends ExternalResource {
+    private static final Logger logger = LoggerFactory.getLogger(ApmServerRule.class);
+
+    private final Queue<String> receivedIntakeEvents = new ConcurrentLinkedQueue<>();
+    private final MockWebServer apmServer = newApmServer(receivedIntakeEvents);
+
+    private static MockWebServer newApmServer(Queue<String> receivedIntakeEvents) {
+        MockWebServer apmServer = new MockWebServer();
+        apmServer.setDispatcher(new ApmDispatcher(receivedIntakeEvents));
```

Comment:
currently, code eagerly calls `getPort()` due to some lifecycle conflation. We can't set the dispatcher after `getPort()` is called.